src/main/java/carparks/parking/ParkingService.java

```diff
@@ -0,0 +1,190 @@
+package carparks.parking;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ParkingService {
+
+	private ParkingDao parkingDao;
+	private FeeCalculator feeCalculator;
+	private Optional<List<CurrencyConverter>> converters;
+
+	@Autowired(required=false)
+	public ParkingService(ParkingDao parkingDao, FeeCalculator feeCalculator, Optional<List<CurrencyConverter>> converters) {
+		this.parkingDao = parkingDao;
+		this.feeCalculator = feeCalculator;
+		this.converters = converters;
+	}
+
+	public ParkingEntryDto tryToCreateParkingEntry(String plateNumber, FeeType feeType) {
+
+		List<ParkingEntry> entries = getStartedSessionsListByPlateNumber(plateNumber);
+
+		if (entries.size() > 0) {
+
+			throw new IllegalStateException("This vehicle has already started a parking session");
+
+		} else {
+
+			ParkingEntry newEntry = new ParkingEntry(plateNumber, feeType);
+			parkingDao.save(newEntry);
+			return new ParkingEntryDto(plateNumber, new BigDecimal(0), SessionStatus.IN_PROGRESS);
+		}
+	}
+
+	public ParkingEntryDto tryToFinishParkingSession(String plateNumber) {
+
+		List<ParkingEntry> entries = getStartedSessionsListByPlateNumber(plateNumber);
+
+		if (entries.size() == 0) {
+
+			throw new IllegalStateException("This vehicle hasn not started a parking session");
+
+		} else {
+
+			ParkingEntry updatedEntry = entries.get(0);
+			updatedEntry.setFinish(LocalDateTime.now());
+			updatedEntry.setFee(feeCalculator.calculateFee(updatedEntry));
+			parkingDao.save(updatedEntry);
+			return new ParkingEntryDto(plateNumber, new BigDecimal(1), SessionStatus.FINISHED);
+
+		}
+
+	}
+
+	public ParkingEntryDto getEntry(String plateNumber, String currency){
+
+		List<ParkingEntry> entriesStarted = getStartedSessionsListByPlateNumber(plateNumber);
+
+		if (entriesStarted.size() > 0) {
+			BigDecimal feeSoFar = feeCalculator.calculateFee(entriesStarted.get(0));
+			
+			if (!currency.equals("PLN")) {
+				feeSoFar = getConvertedValue(feeSoFar, currency);
+			}
+			
+			return new ParkingEntryDto(plateNumber, feeSoFar, SessionStatus.IN_PROGRESS);
+		}
+		
+		ParkingEntry mostRecent = getMostRecentFinishedSession(plateNumber);
+		BigDecimal fee = mostRecent.getFee();
+		if (!currency.equals("PLN")) {
```

Comment:
A jeśli w którymś miejscu byś napisał "PNL", to czy zauważyłbyś to? Wiesz w jaki sposób mogłeś ustrzec się przed takimi pomyłkami?

Reply:
Należy wyciągać wartości tego typu do stałych, np do pliku .properties.

Reply:
Na tym etapie wystarczyłaby stała w kodzie, ale ogólnie zgadza się.